### ISSUE TITLE
azurite blob store returns 400 Bad Request when accessed over a DNS hostname

### ISSUE THREAD

**Issue by akohring** [Azure/Azurite#1039]
### Which service(blob, file, queue, table) does this issue concern?
blob

### Which version of the Azurite was used?
3.14.2

### Where do you get Azurite? (npm, DockerHub, NuGet, Visual Studio Code Extension)
DockerHub

### What's the Node.js version?
N/A

### What problem was encountered?
400 Bad Request is returned when accessing blobs over a network using a DNS hostname rather than IP.

### Steps to reproduce the issue?

- Add the following entry to /etc/hosts: 127.0.0.1	foo.bar.com
- Run azurite locally
- curl -v -XPUT -H "Authorization: SharedKey devstoreaccount1:xxx" http://foo.bar.com:10000/devstoreaccount1/container?restype=container

### Have you found a mitigation/solution?
Downgrade to v3.14.0

**Comment by dwsteele** [comment]
We saw a similar issue and downgraded to 3.14.0 to fix the issue: https://github.com/pgbackrest/pgbackrest/commit/c38d6926d6c9aa01b895a28e66fc0aa6965350a3

In our case it looks like there is only a problem when the hostname is multi-part, i.e. `azure` as a host name seems OK but `azure.example.com` is not. In our doc builds we actually use `blob.core.windows.net` in `/etc/hosts` to "fake" real azure.

**Comment by XiaoningLiu** [comment]
Sorry for the inconvenience due to the change. That's a feature contributed in #911.

After the feature change, there will be 2 modes accessing to Azurite:

### IP Style

Storage account name is in 1st part of URL path. For example,

` http://127.0.0.1:10000/devstoreaccount1/container`

` http://localhost:10000/devstoreaccount1/<resource-path>`

` http://<single part hostname defined in hosts file>:10000/devstoreaccount1/<resource-path>`

### Production Style

For production style URL, the account name is in the first part of URL host.
This aligns with Azure Storage production to provide a unified experience between Azure Storage production with Azurite.

For example,

` http://accountname.<multi part host suffix>:10000/<resource-path>`

After map devstoreaccount1.blob.core.windows.net to 127.0.0.1 in hosts file:

` http://devstoreaccount1.blob.core.windows.net:10000/<resource-path>`

**Comment by dwsteele** [comment]
I think it would be a good idea to mention breaking changes in the release notes as well.

Honestly, I think this would have been better implemented as a server switch. Switching based on the host name in the request is a bit chaotic and means that clients/users need to understand the Azurite logic in order to get it right. If Azurite tweaks the logic it means more breakage for clients/users. It appears that the regexp does not support IPv6 so those requests are still going to use host-style URIs.

**Comment by dotariel** [comment]
I agree with @dwsteele. This feels like an unnecessary complexity.

**Comment by pgguru** [comment]
Seems like a compatibility issue that would have been useful to turn on/off (with default off for existing installs) to prevent disruption.  Allowing this to be configurable by choice just seems like a better design decision.

**Comment by m-jovanovic** [comment]
Just spent a day on this, without being able to figure out what the issue was... Thank god I found this thread.

Would be really nice if something like this was outlined in the documentation. I just started out with Azurite and had a rough time trying to integrate it with my ASP.NET Core application all running using docker-compose.

For anyone else running into this issue this is the connection string that worked for me:
`"UseDevelopmentStorage=true;DevelopmentStorageProxyUri=http://blob;"`

`blob` is the name of the container, in my case.

**Comment by blueww** [comment]
Raise a PR to update readme with the DNS name change: https://github.com/Azure/Azurite/pull/1064

**Comment by blueww** [comment]
PR #1064 is already merged, by add the limitation of [product style URI ](https://github.com/Azure/Azurite#production-style-url)to readme.

**Comment by justinmchase** [comment]
Please don't publish breaking changes in patch version bumps :(

I had multiple people out of commission for a couple days trying to track down why this was happening for only some people...

Side note using `imagePullPolicy: IfNotPresent` in conjunction with `:latest` image tags is a _baaad_ idea.

**Comment by dotariel** [comment]
I agree with @justinmchase . I would not use this in a production setting so applying production-like awareness seems like an unnecessary complexity.

**Comment by blueww** [comment]
PR https://github.com/Azure/Azurite/pull/1112 has been merged to fix this issue, and the fix will be in the next release.
With the fix, when start azurite with `--disableProductStyleUrl`, Azurite will parse request Uri in the same way as before (won't get account name from request Uri host, but from the 1st path segment of request Uri.)
